### ISSUE TITLE
Extract method breaks compilation by introducing naming conflicts

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#22520]
Extract method refactoring can't handle out variables from deconstruction declarations and declaration patterns. It also doesn't report conflicts for such cases.

**Version Used**: 
```
Microsoft Visual Studio Professional 2017 
Version 15.3.5
VisualStudio.15.Release/15.3.5+26730.16
Microsoft .NET Framework
Version 4.7.02046
```

**Steps to Reproduce**:

Use `extract method` refactoring to extract first 4 lines (either all of them at once or line by line) of `Main` method in the following code:
```cs
class C
{
  static void Main(string[] args)
  {
    var (a, b) = (0, 0);
    _ = a is var c;
    if (a is int d || (d = 0) > 0) { }
    if (a is int e) { } else e = 0;

    Console.WriteLine(a);
    Console.WriteLine(c);
    Console.WriteLine(d);
    Console.WriteLine(e);
  }
}
```

**Expected Behavior**:
Either conflicts should be reported or refactoring should produce code without compilation errors

**Actual Behavior**:
No conflicts. Refactoring creates this code:
```cs
using System;

class C
{
  static void Main(string[] args)
  {
    int a, c, d, e;
    NewMethod(out a, out c, out d, out e);

    Console.WriteLine(a);
    Console.WriteLine(c);
    Console.WriteLine(d);
    Console.WriteLine(e);
  }

  private static void NewMethod(out int a, out int c, out int d, out int e)
  {
    var (a, b) = (0, 0);
    _ = a is var c;
    if (a is int d || (d = 0) > 0) { }
    if (a is int e) { } else e = 0;
  }
}
```

Roslyn output for the generated code:
```
Error CS0177: The out parameter 'a' must be assigned to before control leaves the current method 
Error CS0177: The out parameter 'c' must be assigned to before control leaves the current method 
Error CS0177: The out parameter 'd' must be assigned to before control leaves the current method 
Error CS0177: The out parameter 'e' must be assigned to before control leaves the current method 
Error CS0136: A local or parameter named 'a' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
Error CS0136: A local or parameter named 'c' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
Error CS0136: A local or parameter named 'd' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
Error CS0136: A local or parameter named 'e' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
```
